docs/source/development/eep-01-pytrees.rst

```diff
@@ -0,0 +1,848 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The second solution is a dictionary of pytrees the keys are the columns of the current
+summary but probably in plural, i.e. "values", "standard_errors", "p-values", ...;
+
+Each value is a pytree with the exact same structure as ``params``. If this pytree
+contains DataFrames with ``"value"`` column, only that column is updated. i.e. standard
+errors would be accessed via ``summary["standard_errors"]["my_df"]["value"]``.
+
+
+Representation of covariance matrices
+-------------------------------------
+
+A covariance matrix is a two dimensional extension of a ``params`` pytree. We could
+theoretically handle it exactly the same way as Jacobians. However, this would not be
+useful for statistical tests and visualization if it contains more than 2 dimensional
+arrays (as the Jacobian example does).
+
+We thus propose to have two possible formats in which covariance matrices can be
+returned:
+
+1. The pytree variant described in the above Jacobian example. This will be useful
+   to look at sub-matrices of the full covariance matrix as long as the ``params``
+   pytree only contains one dimensional arrays, Series and DataFrames with ``"value"``
+   columns.
+2. A DataFrame containing the covariance matrix of the flattened parameter vector. The
+   index and columns of the DataFrames can be constructed from the ``leaf_names``
+   function in ``pybaum``. We could also triviall add a function there that constructs
+   an index that is easier to work with for selecting elements and let the user choose
+   between the two versions.
+
+The function that maps from the flat version (which would be calculated internally) to
+the pytree version is the same as we need for numerical derivatives. The inverse of that
+function is probably not too difficult to implement and can also be useful for
+derivatives.
+
+
+params
+------
+
+Everything that can be used as ``params`` in optimization and differentiation can
+also be used as ``params`` in estimation. The registries used in pytree functions
+are identical.
+
+
+ML specific aspects of pytrees
+==============================
+
+The output of the log likelihood functions is a dictionary with the entries:
+
+- ``"value"``: a scalar float
+- ``"contributions"``: a 1d numpy array or pandas.Series
+
+Moreover, there can be arbitrary additional entries.
+
+The only change is that ``"contributions"`` can now be any estimagic pytree.
+
+
+MSM specific aspects of pytrees
+===============================
+
+
+Valid formats of empirical and simulated moments
+------------------------------------------------
+
+There are three types of moments in MSM estimation:
+
+- ``empirical moments``
+- The output of ``simulate_moments``
+- The output of ``calculate_moments``, needed to get a moments covariance matrix.
+
+We propose that moments can be stored as any valid estimagic pytree but of course all
+three types of moments have to be aligned, i.e. be stored in a tree of the same
```

Comment:
Is there a utility for checking this? I saw something on equality comparisons, no time anymore to look it up right now, but I am not sure anymore whether that referred to equality of structures or equality of values in leafs.

Reply:
Yes, can be checked. Added it.